services/src/main/java/org/keycloak/services/resources/ThemeResource.java

```diff
@@ -100,19 +105,43 @@ public Response getResource(@PathParam("version") String version, @PathParam("th
             boolean hasContentHash = theme.hasContentHash(path);
 
             if (Profile.isFeatureEnabled(Profile.Feature.ROLLING_UPDATES_V2)) {
+                // A simpler way to check for encoded URL characters would be to retrieve the raw values.
+                // Unfortunately, RESTEasy doesn't support this, and UrlInfo will throw an IllegalArgumentException.
+
+                String base = uriInfo.getBaseUri().getPath();
+                base = base.substring(0, base.length() - 1);
+                if (!uriInfo.getRequestUri().toURL().getPath().startsWith(base + UriBuilder.fromResource(ThemeResource.class)
+                        .path("/{version}/").build(version).getPath())) {
+                    log.debugf("No URL encoding should be necessary for the version, returning a 404: %s", uriInfo.getRequestUri().getPath());
```

Comment:
My understanding is that RestEasy will always return the `version` in decoded form. I'm struggling to understand scenarios where we need this check, can you provide an example?

Reply:
I am trying to prevent situations where someone passes in strange versions in, which then lead to open an open or half-open redirect.

Imagine a version like `myversion/../../auth/realm/xxx` which is then url-encoded as `myversion%2F..%2F..%2Fauth%2Frealm%2Fxxx` and passed in as the `version` URL fragment. 

So I'd say everything that is encoded is possibly someone trying to exploit it. [And as a version, we only use characters and numbers at the moment.](https://github.com/keycloak/keycloak/blob/7cff8572387e6bbfca160aa60e9a8191d22e5c85/model/jpa/src/main/java/org/keycloak/models/jpa/MigrationModelAdapter.java#L87-L91) I could add that as a regex check in addition, or as a replacement. 

Still, IMHO we need the encoding check for the path.

Reply:
Yes and no: For that branch, it is. But then there is also the part in line 148, and the fallthrough where there is a content hash and the resource is returned to the caller starting in line 156. For that item, I also don't want to accept just any version.